cpp/src/group_public_key_context.cpp

```diff
@@ -0,0 +1,109 @@
+    if (XTT_RETURN_SUCCESS != ctor_ret) {
+        throw std::runtime_error("Error encountered in LRSW group public key context constructor");
+    }
+}
+
+struct xtt_group_public_key_context* group_public_key_context_lrsw::get()
+{
+    return &gpk_ctx_;
+}
+
+const struct xtt_group_public_key_context* group_public_key_context_lrsw::get() const
+{
+    return &gpk_ctx_;
+}
+
+group_public_key_context_lrsw::group_public_key_context_lrsw(const std::vector<unsigned char>& basename,
+                                                             const std::vector<unsigned char>& serialized_lrsw_gpk)
+{
+    if (MAX_BASENAME_LENGTH < basename.size()) {
+        throw std::runtime_error("Bad serialized basename value passed to LRSW group public key context constructor");
+    }
+
+    if (sizeof(xtt_daa_group_pub_key_lrsw) != serialized_lrsw_gpk.size()) {
+        throw std::runtime_error("Bad serialized GPK value passed to LRSW group public key context constructor");
+    }
+
+    xtt_return_code_type ctor_ret =
+        xtt_initialize_group_public_key_context_lrsw(&gpk_ctx_,
+                                                     basename.data(),
+                                                     basename.size(),
+                                                     reinterpret_cast<const xtt_daa_group_pub_key_lrsw*>(serialized_lrsw_gpk.data()));
+    if (XTT_RETURN_SUCCESS != ctor_ret) {
+        throw std::runtime_error("Error encountered in LRSW group public key context constructor");
+    }
+}
+
+group_public_key_context_lrsw::group_public_key_context_lrsw(const std::string& as_text_basename,
+                                                             const std::string& serialized_as_text_lrsw_gpk)
+    : group_public_key_context_lrsw(text_to_binary(as_text_basename),
+                                    text_to_binary(serialized_as_text_lrsw_gpk))
+{
+}
+
+std::pair<std::vector<unsigned char>, std::vector<unsigned char>> group_public_key_context_lrsw::serialize() const
+{
+    return std::make_pair(std::vector<unsigned char>(gpk_ctx_.basename,
+                                                     gpk_ctx_.basename+std::min<std::size_t>(gpk_ctx_.basename_length, MAX_BASENAME_LENGTH)),
+                          std::vector<unsigned char>(gpk_ctx_.gpk.lrsw.data,
+                                                     gpk_ctx_.gpk.lrsw.data+sizeof(xtt_daa_group_pub_key_lrsw)));
+}
+
+std::pair<std::string, std::string> group_public_key_context_lrsw::serialize_to_text() const
+{
+    return std::make_pair(binary_to_text(gpk_ctx_.basename, std::min<std::size_t>(gpk_ctx_.basename_length, MAX_BASENAME_LENGTH)),
+                          binary_to_text(gpk_ctx_.gpk.lrsw.data, sizeof(xtt_daa_group_pub_key_lrsw)));
+}
+
+group_identity group_public_key_context_lrsw::gid_from_sha256() const
+{
+    uint16_t hash_len;
+    std::vector<unsigned char> raw_gid(sizeof(xtt_sha256));
```

Comment:
This logic should probably be in an `xtt` function `group_public_key_context`?  I had to reimplement this hashing in Python too.  And it'd be nice if `xtt_crypto_hash_sha256` didn't need to be part of the public API.